Simple gripper demo

amcastro-tri opened the pull request:
Simple gripper demo to test the performance of our contact models.
Thanks @SeanCurtis-TRI and @rcory for providing the initial version of this demo.

This demo helped detect a bug in MBP which this PR also fixes.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8873)
<!-- Reviewable:end -->

amcastro-tri commented:
+@rcory for feature review please.
+@seancurtis-tri for platform review please.

I assigned you guys for review since you are very familiar with this test.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8873#-:-LDHvhEWCXhUjHdtDHZ5:b-ya7off)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
First pass 

1. There are a number of issues you inherited. However, since you're pushing it into master, the burden of Drake standards falls on you.  Sorry about that (since I'm one of the guilty parties that gave you some shady code).
2. You've got some CI issues

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[examples/simple_gripper/run_simple_gripper.sh, line 3 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI2nw52SRRZQE8sdTE:-LDI2nw52SRRZQE8sdTF:b-nomhnz) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/run_simple_gripper.sh#L3)):*
> ```Shell
> #!/usr/bin/env bash
> 
> # Helper script to run the example from the root drake directory. It changes the
> ```

BTW I the presence of a shell script requires a README.md file. That's the natural, first place for someone to look for execution information. In fact, the comment in this file even *refers* to a non-existing README.md file.

---

*[examples/simple_gripper/run_simple_gripper.sh, line 8 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI3JQ00SwAn9uKFGCV:-LDI3JQ00SwAn9uKFGCW:b6eyj6i) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/run_simple_gripper.sh#L8)):*
> ```Shell
> # See README.md for a description of the demo and simulation parameters.
> 
> # We need to get into the .runfiles directory in order for sdformat to find root.sdf. See #7874.
> ```

BTW line width

---

*[examples/simple_gripper/run_simple_gripper.sh, line 23 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI3Gp8-8eGp2hf9yYd:-LDI3Gp93lAUUOpmQW4q:b6eyj6i) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/run_simple_gripper.sh#L23)):*
> ```Shell
> 
> # Note on grip width:
> # With --finger_width=0.09415, the gripper just barely touches the mug and it starts falling
> ```

BTW line width

---

*[examples/simple_gripper/simple_gripper.cc, line 57 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI6Q0i7jWYxraNhVNN:-LDI6Q0i7jWYxraNhVNO:b-qjqkfc) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L57)):*
> ```cpp
> using drake::systems::SemiExplicitEulerIntegrator;
> using drake::systems::Sine;
> ```

BTW The command-line docs have erratic units. Units for all, please.  And, for preference, with the same presentation. (e.g., not sometimes "(m)" and sometimes "in meters").

---

*[examples/simple_gripper/simple_gripper.cc, line 68 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI5rKr4Ig19fftcGU_:-LDI5rKr4Ig19fftcGUa:b-qdph0q) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L68)):*
> ```cpp
>               "The initial distance between the gripper fingers.");
> 
> // Integration paramters:
> ```

BTW s/paramters/parameters

---

*[examples/simple_gripper/simple_gripper.cc, line 75 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI5wCq0FqDff5X0SAe:-LDI5wCq0FqDff5X0SAf:b-406tsr) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L75)):*
> ```cpp
> DEFINE_double(max_time_step, 1.0e-3,
>               "Time maximum step used for the integrators. "
>               "If negative, a value based on patemeter penetration_allowance "
> ```

BTW s/patemeter/parameter

---

*[examples/simple_gripper/simple_gripper.cc, line 78 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI6D9HDvKQQH-2vazZ:-LDI6D9HDvKQQH-2vaz_:b9af83t) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L78)):*
> ```cpp
>               "is used.");
> DEFINE_double(accuracy, 1.0e-2, "Sets the simulation accuracy for variable step"
>     "size integrators with error control.");
> ```

BTW indentation

---

*[examples/simple_gripper/simple_gripper.cc, line 82 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI6FPSC5mQzmmv3BUa:-LDI6FPSC5mQzmmv3BUb:b-car63c) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L82)):*
> ```cpp
> // Contact parameters
> DEFINE_double(penetration_allowance, 1.0e-2,
>               "Penetration allowance, in meters");
> ```

BTW Like with `target_realtime_rate`, probably worth noting the documentation that would explain this value.

---

*[examples/simple_gripper/simple_gripper.cc, line 120 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI8IFD3oqnONozqUSD:-LDI8IFD3oqnONozqUSE:b-uo0nkq) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L120)):*
> ```cpp
> void AddGripperPads(MultibodyPlant<double>* plant,
>                     SceneGraph<double>* scene_graph,
>                     const double pad_offset, const Body<double>& finger) {
> ```

BTW `pad_offset` desperately needs to be more formally documented. And probably have a better name that represents its mathematical role more clearly.

---

*[examples/simple_gripper/simple_gripper.cc, line 126 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI7n3I4A1mUv833gzL:-LDI7n3JBte_Ari7cZrT:b-fxs7xo) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L126)):*
> ```cpp
> 
>   Vector3d p_FSo;  // Position of the sphere frame S in the finger frame F.
>   // The gripper frame is defined as:
> ```

BTW I'd recommend throwing the symbol `G` in here so it doesn't look like a typo in reinterpreting frame `F`. Also (as per subsequent note), you should also give a definition for `X_GF`.

---

*[examples/simple_gripper/simple_gripper.cc, line 133 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI7Pqa96P1ZE4-DuMt:-LDI7Pqa96P1ZE4-DuMu:b-f8l21z) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L133)):*
> ```cpp
>     p_FSo(0) = pad_offset;  // Offset from the center of the gripper.
>     p_FSo(1) =
>         std::cos(d_theta * i + sample_rotation) * kPadMajorRadius + 0.0265;
> ```

BTW The magic number 0.0265 deserves an explanation.

---

*[examples/simple_gripper/simple_gripper.cc, line 136 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI7bueCqa_KFkLdfP4:-LDI7bueCqa_KFkLdfP5:b3pyylj) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L136)):*
> ```cpp
>     p_FSo(2) = std::sin(d_theta * i + sample_rotation) * kPadMajorRadius;
> 
>     // Pose of the sphere frame S in the gripper frame G.
> ```

BTW this doesn't document `X_GF`. So, going from `p_FSo` to `X_GS` skips a step.

---

*[examples/simple_gripper/simple_gripper.cc, line 174 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI8ZXL6pc5rh1jU82l:-LDI8ZXL6pc5rh1jU82m:bgdyd2d) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L174)):*
> ```cpp
>   const double pad_offset = 0.0046;
>   if (FLAGS_gripper_force == 0) {
>     // We then fix everything to the right finger and leave the left finger
> ```

BTW This comment feels out of place -- it's about leaving things free and forces -- but the code is all about pad placement.

---

*[examples/simple_gripper/simple_gripper.cc, line 232 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI96HfExoVfuGJ3ALH:-LDI96HfExoVfuGJ3ALI:bj4c2zt) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L232)):*
> ```cpp
> 
>   // Sinusoidal force input. We want the gripper to follow a trajectory of the
>   // form x(t) = X0 * sin(ω⋅t). By differentiating once, we get the velocity
> ```

BTW X0 suggests initial condition, but this is amplitude. Is there a better symbol?  (E.g., A)?  Or are we really missing the term `x(t) = x(0) + A * sin(ω⋅t)`?

---

*[examples/simple_gripper/simple_gripper.cc, line 233 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDIAf2V1nUyvJU2hJ42:-LDIAf2V1nUyvJU2hJ43:bwulgcv) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L233)):*
> ```cpp
>   // Sinusoidal force input. We want the gripper to follow a trajectory of the
>   // form x(t) = X0 * sin(ω⋅t). By differentiating once, we get the velocity
>   // initial condition, and by differentiating twice, we get the input force we
> ```

BTW '...and by differentiating twice, we __*can compute*__ the input force we need to apply."

(Pedantic, I know. :D)

---

*[examples/simple_gripper/simple_gripper.cc, line 236 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI9Jd6AJ8hJ1PdTRfp:-LDI9Jd6AJ8hJ1PdTRfq:b5pzhnd) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L236)):*
> ```cpp
>   // need to apply.
>   // The mass of the mug is ignored.
>   // TODO(amcastro-tri): add a PD controller to precisely controll the
> ```

BTW s/controll/control

---

*[examples/simple_gripper/simple_gripper.cc, line 245 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDIAqcm4_XVgajdMNBJ:-LDIAqcm4_XVgajdMNBK:b-859ju7) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L245)):*
> ```cpp
>   const double omega = 2 * M_PI * FLAGS_frequency;  // rad/s.
>   const double x0 = FLAGS_amplitude;  // meters.
>   const double f0 = omega * omega * x0 * mass;  // Force amplitude, Newton.
> ```

BTW Why not f = ma? (With a defined two lines down.)

---

*[examples/simple_gripper/simple_gripper.cc, line 251 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI9rwMBr_IsbKee6_P:-LDI9rwMBr_IsbKee6_Q:byegaod) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L251)):*
> ```cpp
> 
>   const Vector2<double> amplitudes(f0, FLAGS_gripper_force);
>   const Vector2<double> frequencies(omega, 0.0);  // 1 Hz
> ```

BTW Comment no longer relevant?

Also, worth noting the trick you're using to get the Sine function to generate a constant?

---

*[examples/simple_gripper/simple_gripper.cc, line 303 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDIBiHt4oMAN3SDpk5w:-LDIBiHu2jd9La9mpedm:b-j8ql58) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L303)):*
> ```cpp
>   const PrismaticJoint<double>& y_translate_joint =
>       plant.GetJointByName<PrismaticJoint>("y_translate_joint");
>   y_translate_joint.set_translation(&plant_context, 0.015);
> ```

BTW Magic number: 0.015

---

*[examples/simple_gripper/simple_gripper.cc, line 334 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDIBuDG84Y-wOpsR1UP:-LDIBuDG84Y-wOpsR1UQ:bqiwxjn) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L334)):*
> ```cpp
>     integrator->set_target_accuracy(FLAGS_accuracy);
> 
>   simulator.set_publish_every_time_step(true);
> ```

BTW A note indicating why you're publishing *every* time step rather than visualizing at 60 Hz?

---

*[examples/simple_gripper/simple_gripper.sdf, line 3 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI4KiX0spEOS38BhWA:-LDI4KiX0spEOS38BhWB:boctf3f) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.sdf#L3)):*
> ```
> <?xml version='1.0'?>
> <sdf version='1.6'>
>   <model name='simple_gripper'>
> ```

BTW I'd recommend a comment that explains what is defined here (and what's not -- no collision elements).

---

*[examples/simple_gripper/simple_gripper.sdf, line 4 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI3S6D5kE9xcaY8p9J:-LDI3S6D5kE9xcaY8p9K:b-foe2qs) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.sdf#L4)):*
> ```
> <sdf version='1.6'>
>   <model name='simple_gripper'>
>     <!-- Pose X_WM of the gripper model M in the world frame W. -->
> ```

FYI This would mean more if we had any sense of the gripper's definition relative to the model frame.

---

*[examples/simple_gripper/simple_gripper.sdf, line 128 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI3xbA8gR-v-goGao_:-LDI3xbB1EyZibTVvWDb:b-eonx1i) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.sdf#L128)):*
BTW Not entirely sure about the comment. I think it means:

> Drake attaches an actuator to all joints with a non-zero effort limit. We don't want an actuator on this joint.

(Same on y_translate_joint.)

---

*[examples/simple_gripper/simple_mug.sdf, line 26 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI4mdM3plouh9MK9ni:-LDI4mdM3plouh9MK9nj:b1z0aa5) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_mug.sdf#L26)):*
> ```
>         </geometry>
> 
>         <surface>
> ```

FYI The presence of this tag implies utility. Is this actually consumed? If not, I'd accept a note indicating that it's currently not consumed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8873#-:-LDICFvg8V1rAHd8oW0s:b-rxcjwv)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[examples/simple_gripper/run_simple_gripper.sh, line 1 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDINiaK4DD5DUCJEpMI:-LDNMZXB0Wq5CvCPXdSk:b-hfpv0l) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/run_simple_gripper.sh#L1)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW I'm inclined to suggest not to include this shell script as part of the example. The information regarding the run directory can easily be included in the README.

Also, I don't believe any of the current drake examples provide a shell script to merely provide default gflag values. Default gflag values should be set in the `.cc` file. 

For reference, the only reason I had this file was to avoid `procman`s limited editor window (which made it hard to access gflag variables during development).
</blockquote></details>

removed

---

*[examples/simple_gripper/run_simple_gripper.sh, line 3 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI2nw52SRRZQE8sdTE:-LDNMb2n7o3hlvvmujf1:bb3n5g3) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/run_simple_gripper.sh#L3)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW I the presence of a shell script requires a README.md file. That's the natural, first place for someone to look for execution information. In fact, the comment in this file even *refers* to a non-existing README.md file.
</blockquote></details>

added README

---

*[examples/simple_gripper/run_simple_gripper.sh, line 8 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI3JQ00SwAn9uKFGCV:-LDMuyx59V7JeiQ9nzeq:b-9eayr7) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/run_simple_gripper.sh#L8)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW redundant line (you already explained why you need to do this right above).
</blockquote></details>

I replaced this file with a `README.md`. It was the idea but I completely forgot.

---

*[examples/simple_gripper/simple_gripper.cc, line 75 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI5wCq0FqDff5X0SAe:-LDMx_urCmOvVNb1j9VS:b-miumg1) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L75)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW s/patemeter/parameter
</blockquote></details>

BTW never heard of the patemeters man?

---

*[examples/simple_gripper/simple_gripper.cc, line 82 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI6FPSC5mQzmmv3BUa:-LDMyGGs-_lt66SycQeh:b-ovwaw6) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L82)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Like with `target_realtime_rate`, probably worth noting the documentation that would explain this value.
</blockquote></details>

Good idea/

---

*[examples/simple_gripper/simple_gripper.cc, line 120 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI8IFD3oqnONozqUSD:-LDMzYY70UQmzfU7tp-H:b-ucfknh) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L120)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW the doc header from `Anzu` is below (slightly modified for use here):

```
/// This uses the parameters of the ring (i.e., rubber pad) to add collision
/// geometries to a finger rigid body. The collision geometries, consisting of
/// a set of small spheres, approximates a torus attached to the finger,
/// centered on the tool frame y-axis.
///
/// @param[in] plant the MultiBodyPlant in which to add the pads.
/// @param[in] scene_graph the associated SceneGraph.
/// @param[in] pad_offset the ring offset along the x-axis in the finger
///			  coordinate frame, i.e., how far the ring protrudes from 
///			  the center of the finger.
/// @param[in] finger the Body representing the finger
```
</blockquote></details>

Thanks!

---

*[examples/simple_gripper/simple_gripper.cc, line 245 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDIAqcm4_XVgajdMNBJ:-LDN1yF74cys1dkiADgM:bii0z9l) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L245)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Why not f = ma? (With a defined two lines down.)
</blockquote></details>

ha! don't this kind of things happen to you when you are hacking your way through? 

---

*[examples/simple_gripper/simple_gripper.cc, line 303 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDIBiHt4oMAN3SDpk5w:-LDN3JQ25qU7g-L7RPSE:b-dnzg8r) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L303)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW Magic number: 0.015
</blockquote></details>

leftovers. 

---

*[examples/simple_gripper/simple_gripper.cc, line 340 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDIiDb_5sgXUxzPw2Pc:-LDNIsrCBiUuNUrs1GTF:b-o323wu) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_gripper.cc#L340)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW I've noticed a bunch of `fmt::print` statements. Consider using `drake::log()->debug()` to better control printout info.
</blockquote></details>

BTW. I actually like `fmt::` better for this. My intention was this not to be debug data but useful information to be printed for anyone running this demo.

---

*[examples/simple_gripper/simple_mug.sdf, line 26 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDI4mdM3plouh9MK9ni:-LDNMGRc1kziSwWxLerH:bs0gbri) ([raw file](https://github.com/robotlocomotion/drake/blob/988030c1a2230cb98eeca3509b419b29bd501df0/examples/simple_gripper/simple_mug.sdf#L26)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

FYI The presence of this tag implies utility. Is this actually consumed? If not, I'd accept a note indicating that it's currently not consumed.
</blockquote></details>

yes, that is how we specify friction coefficients now with sdformat. Probably we can customize tags when sdformat allows it (now it doesn't).

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8873)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with a few typos

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[examples/simple_gripper/README.md, line 16 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDOXoh8D4SthB4kc0O2:-LDOXoh8D4SthB4kc0O3:b5ro81g) ([raw file](https://github.com/robotlocomotion/drake/blob/ea38ced4ad701560bb6772eb803942ce103a5bd1/examples/simple_gripper/README.md#L16)):*
> ```Markdown
> `--grip_width` specifies the initial condition for the fingers position.
> 
> The demo also allows to specify a vertical harmonic motion of the gripper to
> ```

BTW "allows to specify" isn't a grammatical phrase. Options:

1. The demo also permits specifying a...
2. The demo allows the user to specify...
3. The demo can move the gripper in a vertical harmonic motion with whose parameters are defined on the command line.

---

*[examples/simple_gripper/README.md, line 20 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDOY8Pz0ShXrFdn_xy9:-LDOY8Pz0ShXrFdn_xyA:b-78q3jn) ([raw file](https://github.com/robotlocomotion/drake/blob/ea38ced4ad701560bb6772eb803942ce103a5bd1/examples/simple_gripper/README.md#L20)):*
> ```Markdown
> motions. The parameters controlling this motion are `--amplitude`, to specify
> the amplitude of the oscillations in meters and, `--frequency` to specify the
> frequency of the oscillations in Hertz. Aditionally, the user might like to
> ```

BTW s/Aditionally/Additionally

---

*[examples/simple_gripper/README.md, line 21 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDOYKlh6tSzViPHIIRS:-LDOYKlh6tSzViPHIIRT:b-o7tcey) ([raw file](https://github.com/robotlocomotion/drake/blob/ea38ced4ad701560bb6772eb803942ce103a5bd1/examples/simple_gripper/README.md#L21)):*
> ```Markdown
> the amplitude of the oscillations in meters and, `--frequency` to specify the
> frequency of the oscillations in Hertz. Aditionally, the user might like to
> change the direction of the oscillations (E.g. to be horizontal or in some other
> ```

BTW not capitalized ("e.g.")

---

*[examples/simple_gripper/README.md, line 47 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDOYW9D4N6-GVMxa7iW:-LDOYW9D4N6-GVMxa7iX:b-v14wvy) ([raw file](https://github.com/robotlocomotion/drake/blob/ea38ced4ad701560bb6772eb803942ce103a5bd1/examples/simple_gripper/README.md#L47)):*
> ```Markdown
> Launch the visualizer
> ```
> bazel-bin/tools/drake_visualizer
> ```

BTW `./bazel-bin/tools/drake_visualizer`  (missing the leading `./`

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8873#-:-LDO_kUUBsv5dqqcG-zX:b-xo3mir)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Sorry guys, but clearly I hadn't run the "clean" version of this example myself. It turns out that when I was playing with it I would remove gravity if I wanted to force oscillations in the vertical axis using this harmonic forcing approach (otherwise the whole thing would start free falling). So I just added a check for vertical oscillations to replicate my hacky experiments. PTAL.

---

Review status: 3 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[examples/simple_gripper/README.md, line 39 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDOluNA-XYuxnoQBazu:-LDRt8Ng8qvwxw7ns3zh:b2r1ev8) ([raw file](https://github.com/robotlocomotion/drake/blob/ea38ced4ad701560bb6772eb803942ce103a5bd1/examples/simple_gripper/README.md#L39)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW I believe this should be `bazel build //examples:simple_gripper`
</blockquote></details>

BTW both work actually. But I'll add the `//` for consistency with the one right above it.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8873#-:-LDSAEOb9t2vftHYWmHY:bcjt8bn)*
<!-- Sent from Reviewable.io -->

rcory commented:
Two things: 

1) Please include this info in the README, so that it is clear that the vertical oscillation case removes gravity. 
2) How do I specify no oscillations (i.e., fixed gripper) with only gravity? Seems that the code throws an exception in it's current form. I think it's still useful to include this option.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---

*[examples/simple_gripper/README.md, line 39 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDOluNA-XYuxnoQBazu:-LDgCvF-EsPDnpJoUnCp:b-kjqd1z) ([raw file](https://github.com/robotlocomotion/drake/blob/ea38ced4ad701560bb6772eb803942ce103a5bd1/examples/simple_gripper/README.md#L39)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW both work actually. But I'll add the `//` for consistency with the one right above it.
</blockquote></details>

BTW I didn't realize that! In any event, since you added the `//` I believe now you also need to add the `:` to the target.

---

*[examples/simple_gripper/README.md, line 16 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDgF0bED0OoqNcKh8Cm:-LDgF0bED0OoqNcKh8Cn:bkato3v) ([raw file](https://github.com/robotlocomotion/drake/blob/33f285d579f5c2e3066456ab180e817b72d4301b/examples/simple_gripper/README.md#L16)):*
> ```Markdown
> `--grip_width` specifies the initial condition for the fingers position.
> 
> The demo also permits specifying a vertical harmonic motion of the gripper to
> ```

BTW how does one specify the no oscillation case (i.e., if I just want to hold the gripper fixed, with gravity)?

---

*[examples/simple_gripper/README.md, line 26 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDgEZ8m5eHyC2xvMaAS:-LDgEZ8nEeTlZwhR6InY:b-au012k) ([raw file](https://github.com/robotlocomotion/drake/blob/33f285d579f5c2e3066456ab180e817b72d4301b/examples/simple_gripper/README.md#L26)):*
> ```Markdown
> model, `simple_gripper.sdf`. Only vertical or horizontal forced motions are
> supported in this example and thus the axis must either be the x-axis or the
> z-axis. Otherwise an exception is thrown communicating this fact.
> ```

BTW It would be good to point out the fact that gravity is removed for the vertical oscillation case.

---

*[examples/simple_gripper/simple_gripper.cc, line 195 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDgFQcz8rF8nk1KSTPo:-LDgFQcz8rF8nk1KSTPp:b1mkh4w) ([raw file](https://github.com/robotlocomotion/drake/blob/33f285d579f5c2e3066456ab180e817b72d4301b/examples/simple_gripper/simple_gripper.cc#L195)):*
> ```cpp
>         -9.81 * Vector3<double>::UnitZ());
>   } else {
>     throw std::runtime_error(
> ```

BTW what if I want no oscillations (i.e., what if this joint doesn't exist or is specified as "Fixed".)? This is still useful use-case to have!

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8873#-:-LDgFsTJ2PVOlmmmTlRx:b-4xbtl9)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
It doesn't throw but you still need to change the translation axis in the SDF file. I added instructions for this particular case.

---

Review status: 5 of 6 files reviewed at latest revision, all discussions resolved.

---

*[examples/simple_gripper/README.md, line 39 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDOluNA-XYuxnoQBazu:-LDgS5uj6T57-yufGu4E:b-olk0da) ([raw file](https://github.com/robotlocomotion/drake/blob/ea38ced4ad701560bb6772eb803942ce103a5bd1/examples/simple_gripper/README.md#L39)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW I didn't realize that! In any event, since you added the `//` I believe now you also need to add the `:` to the target.
</blockquote></details>

BTW nope, not needed.

---

*[examples/simple_gripper/README.md, line 16 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDgF0bED0OoqNcKh8Cm:-LDgW14v6614W4QcvO6Y:b-wjac8t) ([raw file](https://github.com/robotlocomotion/drake/blob/33f285d579f5c2e3066456ab180e817b72d4301b/examples/simple_gripper/README.md#L16)):*
<details><summary><i>Previously, rcory (Rick Cory) wrote…</i></summary><blockquote>

BTW how does one specify the no oscillation case (i.e., if I just want to hold the gripper fixed, with gravity)?
</blockquote></details>

BTW added instructions for this particular case.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8873#-:-LDgW9BHCZ4YhKYIaZvf:b1h6njl)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[examples/simple_gripper/README.md, line 35 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8873#-LDgdh0uDZwkUMeo2maC:-LDgkHbe9jCtX8ZepX1f:b-6ikxe4) ([raw file](https://github.com/robotlocomotion/drake/blob/7dee8cf0c90c3ce79191d267faeb103b227b0639/examples/simple_gripper/README.md#L35)):*
BTW Altenatively:

> ...is set to zero, the fingers will be placed at a fixed `--grip_width` distance apart.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8873)*
<!-- Sent from Reviewable.io -->
